File: leetcode.com/python/340_Longest_Substring_with_At_Most_K_Distinct_Characters.py
# from collections import defaultdict
# class Solution(object):
#     def lengthOfLongestSubstringKDistinct(self, s, k):
#         """
#         :type s: str
#         :type k: int
#         :rtype: int
#         """
#         if not s or len(s) <= 0 or not k or k == 0:
#             return 0
#         charCountDict = defaultdict(int)
#         startIdx = 0
#         longestSubStrLen = 0
#         for endIdx, currentChar in enumerate(s):
#             if len(charCountDict) < k:
#                 charCountDict[currentChar] += 1
#             elif len(charCountDict) == k:
#                 if currentChar in charCountDict:
#                     charCountDict[currentChar] += 1
#                 else:
#                     while len(charCountDict) == k:
#                         charToRemove = s[startIdx]
#                         charCountDict[charToRemove] -= 1
#                         if charCountDict[charToRemove] == 0:
#                             del charCountDict[charToRemove]
#                         startIdx += 1
#                     charCountDict[currentChar] = 1
#             longestSubStrLen = max(longestSubStrLen, endIdx - startIdx + 1)
#         return longestSubStrLen


# My solution during Mock
from collections import defaultdict


class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if len(s) <= 0 or k == 0:
            return 0
        left, right, maxLen = 0, 0, float("-inf")
        frequencies = defaultdict(int)
        while right < len(s):
            frequencies[s[right]] += 1
            while len(frequencies) > k:
                frequencies[s[left]] -= 1
                if frequencies[s[left]] == 0:
                    del frequencies[s[left]]
                left += 1
            maxLen = max(maxLen, right - left + 1)
            right += 1
        return maxLen


"""
length of the longest substring contains at most k distinct characters >> longest substring 

"eceba"

e:1
b:1
"""

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/340_Longest_Substring_with_At_Most_K_Distinct_Characters.swift
import Foundation
class Solution {
    func lengthOfLongestSubstringKDistinct(_ s: String, _ k: Int) -> Int {
        guard k > 0 else {
            return 0
        }
        var sArr = Array(s)
        var counter = [Character:Int]()
        var maxLen = 0
        var (left, right) = (0, 0)
        while right < sArr.count {
            if counter[sArr[right]] != nil || counter.keys.count < k {
                counter[sArr[right], default: 0] += 1
                maxLen = max(maxLen, right - left + 1)
                right += 1
            } else {
                counter[sArr[left], default: 0] -= 1
                if counter[sArr[left]] == 0 {
                    counter.removeValue(forKey: sArr[left])
                }
                left += 1
            }
        }
        return maxLen
    }
}
